_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+
+-   As one should when one sets out for an adventurous journey, we [started](https://blogs.rstudio.com/ai/posts/2021-03-10-forecasting-time-series-with-torch_1/) with an in-depth study of the tools at our disposal: recurrent neural networks (RNNs). We trained a model to predict the very next observation in line, and then, thought of a clever hack: How about we use this for multi-step prediction, feeding back individual predictions in a loop? The result , it turned out, was quite acceptable.
+
+-   [Then](https://blogs.rstudio.com/ai/posts/2021-03-11-forecasting-time-series-with-torch_2/), the adventure really started. We built our first model "natively" for multi-step prediction, relieving the RNN a bit of its workload and involving a second player, a tiny-ish multi-layer perceptron (MLP). Now, it was the MLP's task to project RNN output to several time points in the future. Although results were pretty satisfactory, we didn't stop there.
+
+-   Instead, we [applied](https://blogs.rstudio.com/ai/posts/2021-03-16-forecasting-time-series-with-torch_3/) to numerical time series a technique commonly used in natural language processing (NLP): sequence-to-sequence (*seq2seq*) prediction. While forecast performance was not much different from the previous case, we found the technique to be more intuitively appealing, since it reflects the *causal* relationship between successive forecasts.
+
+Today, we don't replace *seq2seq* by something completely different. Instead, we enrich it, adding in a new component: the so-called *attention* module. Originally introduced around 2014[^1], attention mechanisms have gained enormous traction, so much so that a recent paper title starts out "Attention is Not All You Need"[^2].
+
+[^1]: For important papers see: @BahdanauCB14, @VinyalsKKPSH14, @XuBKCCSZB15.
+
+[^2]: @2021arXiv210303404D, thus replying to @2017arXiv170603762V, whose 2017 paper "went viral" for stating the opposite.
+
+The idea is the following. In the classic encoder-decoder setup, the decoder gets "primed" with an encoder summary just a single time: the time it starts its forecasting loop. From then on, it's on its own. With attention, however, it gets to see the complete sequence of encoder outputs again every time it forecasts a new value. And more than that: Every time, it gets to zoom in on *those* outputs that seem *relevant for the current prediction step*.
+
+We can easily see how this is a useful strategy in translation: In generating the next word, a model will need to know what part of the source sentence to focus on. How much the technique helps with numerical sequences, in contrast, will likely depend on the features of the series in question. Meaning: Even if, in our current case, it should turn out to have little effect, it will still be worthwhile just to have learned about the technique.
```

Comment:
```suggestion
This is a particularly useful strategy in translation: In generating the next word, a model will need to know what part of the source sentence to focus on. How much the technique helps with numerical sequences, in contrast, will likely depend on the features of the series in question. Meaning: Even if, in our current case, it should turn out to have little effect, it will still be worthwhile just to have learned about the technique.
```

Reply:
> Meaning: Even if, in our current case, it should turn out to have little effect, it will still be worthwhile just to have learned about the technique.

Do you need this sentence? It feels like this is a case of you justifying why what you're writing is worthwhile, when I'd assume you wouldn't post it were that not the case.

Reply:
As an aside, the colon-before-a-capitalized-sentence structure is sufficiently rare that I'd try to avoid doing it more than once in a single paragraph.

Reply:
> Do you need this sentence? It feels like this is a case of you justifying why what you're writing is worthwhile, when I'd assume you wouldn't post it were that not the case.

...  ---> ...

> As an aside, the colon-before-a-capitalized-sentence structure is sufficiently rare that I'd try to avoid doing it more than once in a single paragraph.

OK, you said that more than once, time to get around my stubbornness!!